runtime/common/src/paras_registrar.rs

```diff
@@ -0,0 +1,675 @@
+		Debtors: map hasher(twox_64_concat) ParaId => T::AccountId;
+	}
+}
+
+decl_error! {
+	pub enum Error for Module<T: Trait> {
+		/// Parachain already exists.
+		ParaAlreadyExists,
+		/// Invalid parachain ID.
+		InvalidChainId,
+		/// Invalid parathread ID.
+		InvalidThreadId,
+		/// Invalid para code size.
+		CodeTooLarge,
+		/// Invalid para head data size.
+		HeadDataTooLarge,
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: <T as frame_system::Trait>::Origin {
+		type Error = Error<T>;
+
+		#[weight = (5_000_000_000, DispatchClass::Operational)]
+		pub fn register_parathread(
+			origin,
+			id: ParaId,
+			genesis_head: HeadData,
+			validation_code: ValidationCode,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(!Paras::contains_key(id), Error::<T>::ParaAlreadyExists);
+
+			let outgoing = <paras::Module<T>>::outgoing_paras();
+
+			ensure!(outgoing.binary_search(&id).is_err(), Error::<T>::ParaAlreadyExists);
+
+			<T as Trait>::Currency::reserve(&who, T::ParathreadDeposit::get())?;
+			<Debtors<T>>::insert(id, who);
+
+			Paras::insert(id, false);
+
+			let genesis = ParaGenesisArgs {
+				genesis_head,
+				validation_code,
+				parachain: false,
+			};
+
+			<paras::Module<T>>::schedule_para_initialize(id, genesis);
+
+			Ok(())
+		}
+
+		#[weight = (10_000, DispatchClass::Operational)]
+		pub fn register_parachain(
+			origin,
+			id: ParaId,
+			genesis_head: HeadData,
+			validation_code: ValidationCode,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+
+			ensure!(!Paras::contains_key(id), Error::<T>::ParaAlreadyExists);
+
+			let outgoing = <paras::Module<T>>::outgoing_paras();
+
+			ensure!(outgoing.binary_search(&id).is_err(), Error::<T>::ParaAlreadyExists);
+
+			Paras::insert(id, true);
+
+			let genesis = ParaGenesisArgs {
+				genesis_head,
+				validation_code,
+				parachain: true,
+			};
+
+			<paras::Module<T>>::schedule_para_initialize(id, genesis);
+
+
+			Ok(())
+		}
+
+		#[weight = (10_000, DispatchClass::Operational)]
+		pub fn deregister_parachain(origin, id: ParaId) -> DispatchResult {
+			ensure_root(origin)?;
+
+			let is_parachain = Paras::take(id).ok_or(Error::<T>::InvalidChainId)?;
+
+			ensure!(is_parachain, Error::<T>::InvalidChainId);
+
+			<paras::Module<T>>::schedule_para_cleanup(id);
+
+			Ok(())
+		}
+
+		#[weight = (10_000_000, DispatchClass::Operational)]
+		pub fn deregister_parathread(origin) -> DispatchResult {
+			let id = ensure_parachain(<T as Trait>::Origin::from(origin))?;
+
+			let is_parachain = Paras::take(id).ok_or(Error::<T>::InvalidChainId)?;
+
+			ensure!(!is_parachain, Error::<T>::InvalidThreadId);
+
+			let debtor = <Debtors<T>>::take(id);
+			let _ = <T as Trait>::Currency::unreserve(&debtor, T::ParathreadDeposit::get());
+
+			<paras::Module<T>>::schedule_para_cleanup(id);
+
+			Ok(())
+		}
+
+		/// Swap a parachain with another parachain or parathread. The origin must be a `Parachain`.
```

Comment:
What's the reasoning for this limitation? Is it not enough that the types of `ParaId` are not equivalent?

Reply:
This is about the origin of the `Call` - parachains can send messages to the relay chain that are executed as calls to modules. So this requirement means that the parachains themselves must initiate the swap